WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostPublishedSettingsViewModel.kt

```diff
@@ -0,0 +1,130 @@
+        canPublishImmediately = PostUtils.shouldPublishImmediatelyOptionBeAvailable(postModel)
+        postModel?.let {
+            _onPublishedLabelChanged.postValue(postSettingsUtils.getPublishDateLabel(postModel, context))
+        }
+    }
+
+    fun publishNow() {
+        _onPublishedDateChanged.postValue(Calendar.getInstance())
+    }
+
+    fun onTimeSelected(selectedHour: Int, selectedMinute: Int) {
+        this.hour = selectedHour
+        this.minute = selectedMinute
+        val calendar = Calendar.getInstance()
+        calendar.set(year!!, month!!, day!!, hour!!, minute!!)
+        _onPublishedDateChanged.postValue(calendar)
+    }
+
+    fun onDateSelected(year: Int, month: Int, dayOfMonth: Int) {
+        this.year = year
+        this.month = month
+        this.day = dayOfMonth
+        _onDatePicked.postValue(Unit)
+    }
+
+    private fun getCurrentPublishDateAsCalendar(postModel: PostModel): Calendar {
+        val calendar = Calendar.getInstance()
+        val dateCreated = postModel.dateCreated
+        // Set the currently selected time if available
+        if (!TextUtils.isEmpty(dateCreated)) {
+            calendar.time = DateTimeUtils.dateFromIso8601(dateCreated)
+        }
+        return calendar
+    }
+
+    fun updatePost(updatedDate: Calendar, post: PostModel?, context: Context) {
+        post?.let {
+            post.dateCreated = DateTimeUtils.iso8601FromDate(updatedDate.time)
+            val initialPostStatus = PostStatus.fromPost(post)
+            val isPublishDateInTheFuture = PostUtils.isPublishDateInTheFuture(post)
+            var finalPostStatus = initialPostStatus
+            if (initialPostStatus == DRAFT && isPublishDateInTheFuture) {
+                // Posts that are scheduled have a `future` date for REST but their status should be set to `published` as
+                // there is no `future` entry in XML-RPC (see PostStatus in FluxC for more info)
+                finalPostStatus = PUBLISHED
+            } else if (initialPostStatus == PUBLISHED && post.isLocalDraft()) {
+                // if user was changing dates for a local draft (not saved yet), only way to have it set to PUBLISH
+                // is by running into the if case above. So, if they're updating the date again by calling
+                // `updatePublishDate()`, get it back to DRAFT.
+                finalPostStatus = DRAFT
+            } else if (initialPostStatus == SCHEDULED && !isPublishDateInTheFuture) {
+                // if this is a SCHEDULED post and the user is trying to Back-date it now, let's update it to DRAFT.
+                // The other option was to make it published immediately but, let the user actively do that rather than
+                // having the app be smart about it - we don't want to accidentally publish a post.
+                finalPostStatus = DRAFT
+                // show toast only once, when time is shown
+                ToastUtils.showToast(
```

Comment:
Should we move the Toast away from VM, into fragment?

Reply:
totally, I actually did that in the following PR but I'll move the change here 👍